src/protocols/rdp/rdp.c

```diff
@@ -227,10 +227,53 @@ static BOOL rdp_freerdp_authenticate(freerdp* instance, char** username,
 
     rdpContext* context = instance->context;
     guac_client* client = ((rdp_freerdp_context*) context)->client;
-
-    /* Warn if connection is likely to fail due to lack of credentials */
-    guac_client_log(client, GUAC_LOG_INFO,
-            "Authentication requested but username or password not given");
+    guac_rdp_client* rdp_client = (guac_rdp_client*) client->data;
+    guac_rdp_settings* settings = rdp_client->settings;
+    char* params[4] = {};
+    int i = 0;
+    
+    if (settings->username == NULL || strcmp(settings->username, "") == 0) {
+        params[i] = "username";
+        rdp_client->rdp_credential_flags |= GUAC_RDP_CRED_FLAG_USERNAME;
+        i++;
+    }
+    
+    if (settings->password == NULL || strcmp(settings->password, "") == 0) {
+        params[i] = "password";
+        rdp_client->rdp_credential_flags |= GUAC_RDP_CRED_FLAG_PASSWORD;
+        i++;
+    }
+    
+    if (settings->domain == NULL || strcmp(settings->domain, "") == 0) {
+        params[i] = "domain";
+        rdp_client->rdp_credential_flags |= GUAC_RDP_CRED_FLAG_DOMAIN;
+        i++;
+    }
+    
+    /* NULL-terminate the array. */
+    params[i] = NULL;
+    
+    if (i > 0) {
+        /* Lock the client thread. */
+        pthread_mutex_lock(&(rdp_client->rdp_credential_lock));
+        
+        /* Send require params and flush socket. */
+        guac_protocol_send_required(client->socket, (const char**) params);
+        guac_socket_flush(client->socket);
+        
+        /* Wait for condition. */
+        pthread_cond_wait(&(rdp_client->rdp_credential_cond), &(rdp_client->rdp_credential_lock));
```

Comment:
We'll need to be careful here - this will potentially block indefinitely. There are additional cases that aren't yet being handled that we need to consider:

* If the connection is closed while this is waiting, is there anything to ensure it stops waiting? Does the blockage need to stop prior to attempting to free FreeRDP structures?
* This lock and the condition need to be properly destroyed in the free handler. I see some new cleanup within the VNC free handler, but not for RDP.
* `pthread_cond_destroy()` is documented as potentially failing with `EBUSY` if another thread is using the condition with `pthread_cond_wait()`: https://pubs.opengroup.org/onlinepubs/009695399/functions/pthread_cond_destroy.html

There is failsafe code in guacd which will forcibly terminate the process if the connection closes yet fails to clean itself up in a reasonable time, but we should aim to avoid that needing to kick in.

Reply:
Okay - I'll dig around and see what I can find to help figure out how to clean it up.  I added the destroy call for the RDP cleanup handler, but I may need further guidance on how to make sure the `pthread_cond` is not stuck waiting.  I'll take a run at it, though.